inductor: promote half/bfloat16 constant to float for cpu vectorization path

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105440

As scalar path, we should also promote half/bfloat16 constant to float for better accuracy, after this PR, the TIMM ```dm_nfnet``` model amp path can be passed.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

XiaobingSuper committed:
inductor: promote half/bfloat16 constant to float for cpu vectorization path

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: promote half/bfloat16 constant to float for cpu vectorization path"


As scalar code, we should also promote half/bfloat16 constant to float for better accuracy, after this PR, the TIMM ```dm_nfnet``` model amp path can be passed.



[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -2092,6 +2092,10 @@ def store_reduction(name, index, value):
 
             @staticmethod
             def constant(val, dtype):
+                if dtype in (torch.float16, torch.bfloat16):
```
we are not checking float16 here: https://github.com/pytorch/pytorch/pull/105440/files#diff-5ab7b0235e2076a5fc6629ba0b109208940f5b94f5c13babc3e0f87cf4fcec82R2120. Can float16 work well in the test you added?

XiaobingSuper replied:
yes， it can works for my test.

jgong5 commented on test/inductor/test_cpu_repro.py:
```diff
@@ -1968,6 +1968,14 @@ def f(a):
+        def f(x):
+            return torch.ops.aten.mul.Tensor(x, 1.7015043497085571)
+
+        for dtype in [torch.float16, torch.bfloat16]:
+            x = torch.randn(4, 5, dtype=dtype)
+            self.common(f, (x,))
```
add a check for the count of generated vec kernel?

XiaobingSuper replied:
added.

XiaobingSuper committed:
Update on "inductor: promote half/bfloat16 constant to float for cpu vectorization path"


As scalar path, we should also promote half/bfloat16 constant to float for better accuracy, after this PR, the TIMM ```dm_nfnet``` model amp path can be passed.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge

XiaobingSuper commented:
@pytorchbot merge

XiaobingSuper commented:
@pytorchbot revert -m "introduce core dumped when init bfloat16 zero tensor"

XiaobingSuper commented:
@pytorchbot revert -m "introduce core dumped when init bfloat16 zero tensor" -c nosignal
